RATIONALE.md

```diff
@@ -1,5 +1,117 @@
+
+The path of least resistance is a marshalling approach. Almost all parameters
+of size used in scheduler plugins are generated protobuf model types. While not
+performant, it is possible to have the host marshal these objects and unmarshal
+them as needed on the guest. Further optimizations could be made as many of the
+underlying data do not change within a plugin lifecycle.
+
+For example, a function to get a pod could look like this on the guest:
+```webassembly
+(import "k8s.io/api" "pod" (func $get_pod
+  (param $buf i32) (param $buf_limit i32)
+  (result (; pod_len ;) i32)))
+```
+
+The host would marshal the entire pod argument to protobuf like so:
+```go
+func k8sApiPodFn(ctx context.Context, mod wazeroapi.Module, stack []uint64) {
+	buf := uint32(stack[0])
+	bufLimit := bufLimit(stack[1])
+
+	pod := filterArgsFromContext(ctx).pod
+	stack[0] = uint64(marshalIfUnderLimit(mod.Memory(), pod, buf, bufLimit))
+}
+```
+
+The main tradeoff to this approach is performance. The model is very large and
+the default unmarshaller creates a lot of garbage. We have some workarounds
+noted below, and also there are options for mitigation not yet implemented:
+
+* updating heavy objects only when they change
+    * this limits the occurrence count of decode performance
+* tracking of alternative encoding libraries which use protobuf IDL
+    * [polyglot][3] can generate code from protos and might automatically convert
+      protos to its more efficient representation.
+
+## How is `framework.CycleState` implemented in WebAssembly?
+
+`framework.CycleState` is a primarily a key value storage. Plugins store values
+in `PreFilter` or `PreScore` for use afterward. This is similar to Go context,
+except the keys must be strings and the values must implement `StateData`.
+
+While one instance of `framework.CycleState` is re-used for all plugins in a
+scheduling cycle, in practice, `StateData` are not shared. For this reason, we
+implement cycle state in the guest. The only responsibility of the wasm guest
+is to reset any state when `PreFilter` is called. All state is invisible from
+the perspective of the host.
+
+### Why isn't `StateData.Clone` handled in WebAssembly?
+
+`Clone` is a special case for preemption. When all Nodes are filtered out, the
+scheduler attempts to make space via `PostFilter`. Preemption results in a
+[parallel][4] call to `SelectVictimsOnNode` which runs in [parallel][5], and
+removes victims (pods) from `PreFilter` state and `NodeInfo` before calling
+[`RunFilterPluginsWithNominatedPods`][6]. If `StateData` wasn't cloneable, the
+original `StateData` values would be lost.
+
+The current WebAssembly implementation skips `Clone`, for now. Since the
+guest holds the cycle state, the most likely path forward would be to export
+a guest function to save the cycle state and return a state ID to restore it
+with. This function could be called when the host side knows it is in the
```

Comment:
How the host knows it's in preemption, extension point functions are stateless.

And `AddNode` can also be called in filtering, see https://github.com/kubernetes/kubernetes/blob/96d853f4b86b1b44bfb9340e9d05f5fc6b3f4bb9/pkg/scheduler/schedule_one.go#L514

`Filter is called a second time` sounds reasonable at the first glance, but people may have their own preemption plugins, so I'm not quite convinced, let me think more about this...

Reply:
I'm afraid we have to pass the cycle state from host to guest because for guest or host, they didn't know whether to process the logic with a cloned state of a real state, e.g. https://github.com/kubernetes/kubernetes/blob/960830bc660872ff5679f84e676474e3cb536e12/pkg/scheduler/framework/runtime/framework.go#L903-L920

We run `RunFilterPlugins` twice here, once with a cloned state because we added the nominated pods, another time  we just run with the real state.

Reply:
@kerthcet The line of comments you are discussing says this is deferred, which means we are not trying to solve it now. I jotted notes about a potential way to solve it, and I don't think you fully understand it. I have read the same lines you mentioned, but reading lines of code isn't time effective way to test approaches.

The biggest issue in implementing the plugin is the complexity of the logic in the framework which is not well unit tested. Even the comments are complicated and sometimes confusing. I would suggest since you have experience with this, trying to make a unit test upstream or in the repo that can produce a problem you think that the approach I mentioned cannot solve. Then, we get out of conjecture or hypothetical and into something that works or doesn't work. Regardless, Let's try to remember this section of the docs is intentionally skipping this part due to the complexity and also offline advice to skip it for now. If you can skip this part that is mentioned to please skip, you can focus on other parts of the PR which then helps get a baseline we can change as necessary.